File: js/src/instructions/transferInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferInstruction = void 0;
const buffer_1 = require("buffer");
const web3_js_1 = require("@solana/web3.js");
function transferInstruction(nameProgramId, nameAccountKey, newOwnerKey, currentNameOwnerKey, nameClassKey, nameParent, parentOwner) {
    const buffers = [buffer_1.Buffer.from(Int8Array.from([2])), newOwnerKey.toBuffer()];
    const data = buffer_1.Buffer.concat(buffers);
    const keys = [
        {
            pubkey: nameAccountKey,
            isSigner: false,
            isWritable: true,
        },
        {
            pubkey: parentOwner ? parentOwner : currentNameOwnerKey,
            isSigner: true,
            isWritable: false,
        },
    ];
    if (nameClassKey) {
        keys.push({
            pubkey: nameClassKey,
            isSigner: true,
            isWritable: false,
        });
    }
    if (parentOwner && nameParent) {
        if (!nameClassKey) {
            keys.push({
                pubkey: web3_js_1.PublicKey.default,
                isSigner: false,
                isWritable: false,
            });
        }
        keys.push({
            pubkey: nameParent,
            isSigner: false,
            isWritable: false,
        });
    }
    return new web3_js_1.TransactionInstruction({
        keys,
        programId: nameProgramId,
        data,
    });
}
exports.transferInstruction = transferInstruction;

File: python/src/instructions/transfer_instruction.py
from typing import Optional

from solders.instruction import Instruction, AccountMeta
from solders.pubkey import Pubkey


def transfer_instruction(
        name_program_id: Pubkey,
        name_account: Pubkey,
        new_owner: Pubkey,
        current_name_owner: Pubkey,
        name_class: Optional[Pubkey] = None,
        name_parent: Optional[Pubkey] = None,
        parent_owner: Optional[Pubkey] = None
) -> Instruction:
    buffers = [bytes([2]), bytes(new_owner)]

    data = b''.join(buffers)

    accounts = [
        AccountMeta(
            pubkey=name_account,
            is_signer=False,
            is_writable=True
        ),
        AccountMeta(
            pubkey=parent_owner or current_name_owner,
            is_signer=True,
            is_writable=False
        ),
    ]

    if name_class:
        accounts.append(
            AccountMeta(
                pubkey=name_class,
                is_signer=True,
                is_writable=False
            )
        )

    if parent_owner and name_parent:
        accounts.append(
            AccountMeta(
                pubkey=Pubkey.default(),
                is_signer=False,
                is_writable=False
            )
        )
        accounts.append(
            AccountMeta(
                pubkey=name_parent,
                is_signer=False,
                is_writable=False
            )
        )

    return Instruction(
        program_id=name_program_id,
        accounts=accounts,
        data=data
    )